Engines/Wine/QuickScript/GoG Script/script.js

```diff
@@ -3,108 +3,107 @@ include("utils.functions.net.download");
-        if (downloadedFile.endsWith(".exe")) {
-            foundExecutable = downloadedFile;
-        }
-    });
-
-    return foundExecutable;
-}
-
-GogScript.prototype.go = function () {
-    var setupWizard = SetupWizard(InstallationType.APPS, this._name, this.miniature());
-
-    setupWizard.presentation(this._name, this._editor, this._applicationHomepage, this._author);
-
-    this.loginToGog(setupWizard);
-    var setupFile = this.download(setupWizard);
-
-    var wine = new Wine()
-        .wizard(setupWizard)
-        .prefix(this._name, this._wineDistribution, this._wineArchitecture, this._wineVersion)
-        .create()
-        .wait();
-
-    this._preInstall(wine, setupWizard);
-
-    wine.gdiplus();
-    wine.run(setupFile, [], wine.prefixDirectory() + "/drive_c/", true, true);
-
-    this._postInstall(wine, setupWizard);
-
-    this._createShortcut(wine.prefix());
-
-    setupWizard.close();
+class GogScript extends QuickScript {
+    constructor() {
+        super();
+    }
+
+    /**
+     * Sets one setup file name so that the script can fetch it from gog.com
+     * @param {string} setupFileName The setup file name
+     * @returns {GogScript} This
+     */
+    gogSetupFileName(setupFileName) {
+        this._setupFileNames = [setupFileName];
+        return this;
+    }
+
+    /**
+     * Sets the setup file(s) name so that the script can fetch it from gog.com
+     * @param {string[]} setupFileNames The setup file name(s)
+     * @returns {GogScript} This
+     */
+    gogSetupFileNames(setupFileNames) {
+        this._setupFileNames = setupFileNames;
+        return this;
+    }
+
+    /**
+     * Presents a Gog.com login window to the user, login to its account and return a token that can be used later.
+     * Stores the tocken in a parameter
+     * @param {SetupWizard} setupWizard The setupWizard to use
+     * @returns {GogScript} This
+     */
+    loginToGog(setupWizard) {
+        const browserWindow = setupWizard.createBrowser(tr("Please login to your GoG.com account so that we can download the game for you:"));
+        browserWindow.goToUrl("https://auth.gog.com/auth?client_id=46899977096215655&redirect_uri=https%3A%2F%2Fembed.gog.com%2Fon_login_success%3Forigin%3Dclient&response_type=code&layout=client2");
+        browserWindow.waitForUrl("https://embed.gog.com/*");
+
+        const currentUrl = browserWindow.getCurrentUrl();
+        const code = currentUrl.split("code=")[1].split("&")[0];
+
+        const tokenUrl = "https://auth.gog.com/token?client_id=46899977096215655&client_secret=9d85c43b1482497dbbce61f6e4aa173a433796eeae2ca8c5f6129f2dc4de46d9&grant_type=authorization_code&code=" + code + "&redirect_uri=https%3A%2F%2Fembed.gog.com%2Fon_login_success%3Forigin%3Dclient";
+        this._token = new Downloader().url(tokenUrl).json();
+        return this;
+    }
+
+    _downloadSetupFile(setupWizard, setupFileName, tmpDirectory) {
```

Comment:
I don't think we need to name our "private" methods with a prefix `_`, because it can't block someone from using it if he really wants to. If we want to make it clear that the method is intended as a private method I think we should add it as a comment before the method, something like:

```javascript
/**
 * *** INTERNAL API ***
 * ...
 */
```

Reply:
I think we do have a private keyword in JS class. I suggest we use it.

Reply:
Are you sure JS contains a private keyword? I'm not aware of such a keyword.

Reply:
Can you open a new issue to track this?

Reply:
See #1022